### PR TITLE
[Event Hubs] Buffered Producer Flush and Clear

### PR DESCRIPTION
**Pull request by jsquire**
# Summary

The focus of these changes is to implement the flush and clear behavior for the `EventHubBufferedProducerClient`.  Riding along as additional minor tweaks and smoothing for the existing client surface.

Pending work:
- Automatic partition assignment
- Partition Key hashing
- Live tests

# Related and References

- [ EventHubBufferedProducerClient: Flush and Clear (#23482)](https://github.com/Azure/azure-sdk-for-net/issues/23482)

### CODE REVIEW

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Diagnostics/EventHubsEventSource.cs [hunk 1]
```diff
@@ -1387,7 +1387,7 @@ public virtual void AmqpManagementLinkCreateComplete(string eventHubName)
         }
 
         /// <summary>
-        ///   Indicates that an <see cref="EventHubBufferedProducerClient" /> instance is about to begin processing events.
+        ///   Indicates that a <see cref="EventHubBufferedProducerClient" /> instance is about to begin processing events.
```
Comment (christothes) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Diagnostics/EventHubsEventSource.cs:1390]: Isn't "an" correct here?

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs [hunk 2]
```diff
@@ -1216,18 +1205,44 @@ public virtual async ValueTask DisposeAsync()
...
-            //  NOTE:
-            //    This method is currently just a stub; full functionality will be
-            //    added in a later set of changes.
-            // ======================================================================
+            var operationId = GenerateOperationId();
+
+            try
+            {
+                Logger.BufferedProducerClearStart(Identifier, EventHubName, operationId);
+                cancellationToken.ThrowIfCancellationRequested();
```
Comment (christothes) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs:1219]: nit: could we move this to the top of the foreach and eliminate the one below at the bottom of it?

**Review by christothes — APPROVED** [review]

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Diagnostics/EventHubsEventSource.cs (hunk 1 above)
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Diagnostics/EventHubsEventSource.cs:1390]: hrmm... yeah.   Looks like I was a bit more permissive with the find/replace than intended.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs (hunk 2 above)
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs:1219]: Yes, yes we could.   That would also eliminate the need to poll the token in the loop as well.

Comment (christothes) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs:1219]: Cool - IIRC, there was one other place with this same pattern.

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs:1219]: Yeah, `FlushInternalAsync` had similar.   I cleaned that up as well.

**Pull request merged**